Improve README for `<LinkControl />` with historical context and props docs.

getdave opened the pull request:
## Description
The current docs for `LinkControl` don't provide any context about how/why it was created. This PR aims to provide this plus generally improve the docs with more detail around `props` and expected value types.

## Todo

- [x] Document [all things suggested by @aduth](https://github.com/WordPress/gutenberg/pull/19462#issuecomment-574732569). 
- [x] Match Types against TS/JS docblocks
- [x] Double check `prop` documentation is accurate.

## How has this been tested?

You can test this by:

* Reading the `LinkControl` source code and comparing JSDoc and TSDoc blocks against README.
* Checking for grammar/spelling errors.
* Checking for areas which don't make sense or could be made clearer.
* Ensuring the historical context matches your understanding of how the component evolved.
* Ensuring nothing in the docs is misleading nor implies other components will be deprecated or superseded.
* Comparing against `master` to see if (subjectively) the docs feel like they are better thanks to the changes suggested here,


## Types of changes
Bug fix (non-breaking change which fixes an issue) 
## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->.

getdave committed:
Scaffold out initial context for Docs

aduth reviewed (commented):
I think this is a good idea from the perspective of providing context for what LinkControl is intended to be, especially in distinguishing itself from other related components.

You and I spoke directly on this, but I am a bit apprehensive to start considering this as a _substitute_ for `URLInput` so much as an additional layer of abstraction on a more minimalistic set of baseline behaviors. Or, at least, I have some hopes that we could still consider `URLInput` to maintain some usefulness on its own. Where I might view "URL Input" as something akin to the `<input type="text">` of a `TextControl` (in this case, the base input upon which `LinkControl` builds additional functionality).

Semantically, I could see it represented as:

- `URLInput` is an input for presenting and managing selection behaviors associated with choosing a URL, optionally from a pool of available candidates.
- `LinkControl` includes this, plus additional behaviors to control how this URL applies to a link (various behaviors such as "opens in new tab", etc).

I don't know if this would be possible, but likewise I'm not sure about committing to verbiage which commits to putting URLInput on the path toward deprecation.

youknowriad commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -1,5 +1,58 @@
+from `core/block-editor` in order to retrieve search suggestions for matching
+`Page` post-type entities.
+
+It is however, possible to provide your own entity search handler via the `fetchSearchSuggestions` prop.
```
I removed this from the component recently since it was not used.

getdave replied:
Which part was not used? Sorry im confused.

youknowriad replied:
This prop was not used in Gutenberg itself, we always relied on the internal implementation.

getdave replied:
Removed 👍

Aurorum commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -1,5 +1,58 @@
+facilitate the implementation of the new UI with the goal of eventually phasing
+out the use of `URLInput` entirely.
+
+Indeed, over time, is is expected that `URLInput` will be depracted in favour of
```
Just spotted what I think is a typo that I thought I'd share. :) 

```suggestion
Indeed, over time, is is expected that `URLInput` will be deprecated in favour of
```

draganescu committed:
fix typo

Co-Authored-By: Aurorum <torres126@clubpenguinmountains.com>

getdave commented:
Going to tackle this on Friday this week and Monday next.

getdave committed:
Remove implications of deprecating URLInput

getdave committed:
Update directly from master to avoid need to rebase.

getdave committed:
Remove ref to legacy `fetchSearchSuggestions` prop

Handles https://github.com/WordPress/gutenberg/pull/19677#discussion_r367296680

getdave committed:
Ensure prop docs match type defs

Also improves documentation grammar and comprehension.

getdave commented:
I've made several adjustments and now believe this is in a better state than when I began.

I'd appreciate a final review.

jeryj commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -70,37 +152,50 @@ Whether to present initial suggestions immediately.
-new suggestion. Suggestions have the following shape:
+As a result of the above, this prop is often used to allow on-the-fly creation of new entities (eg: `Posts`,
+`Pages`) based on the text the user has entered into the link search UI. As an
+example, the Navigation Block uses `createSuggestion` to create Pages on fly
```
"to create Pages on fly" -> "to create Pages on-the-fly"

getdave replied:
Nice spot. Fixed.

jeryj reviewed (changes requested):
Overall I think these docs are really helpful. I especially liked the history and explained reasoning behind *why* this component was built. 

I left a couple of small grammar correction comments. There are also a lot of line-breaks hardcoded in here. I'm guessing it had to do with some kind of copy/paste or auto-formatting? I'd expect they should all just be one long line in the docs instead of having any line breaks.

Other than these minor things, I think the content is great!

getdave committed:
Correct markdown and (wrong) US spelling

getdave commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -1,8 +1,85 @@
+
+## Relationship to `<URLInput>`
+
+As of Gutenberg 7.4, `<LinkControl> became the default link creation interface
```
Fixed.

getdave committed:
Remove line breaks introduced via VSCode add on

getdave commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -1,8 +1,85 @@
+In addition, `<LinkControl>` also allows for on-the-fly creation of links based
+on the **current content of the `<input>` element**. When enabled, a default
+"Create new" search suggestion is appended to all non-URL-like search results.
+When this suggestion is
```
Fixed all these now. Due to VSCode.

getdave committed:
Correct grammar

getdave commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -35,7 +115,9 @@ Default properties include:
-An array of settings objects. Each object will used to render a `ToggleControl` for that setting.
+An array of settings objects associated with a link (for example: a setting to
+determine whether or not the link
+opens in a new tab). Each object will used to render a `ToggleControl` for that setting.
```
Yes. Whoops! Fixed.

getdave committed:
Adds missing word to ensure phrase makes sense.

jeryj commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -1,8 +1,49 @@
+These older UIs tended to make use of two existing components: `URLInput` and `URLPopover`. When a requirement was raised to implement a new UI for hyperlink creation, an assessment of these existing components was undertaken and it was determined that they were too opinionated as to be easily refactored to accommodate the new use cases required by the new UI. Attempting to do so would also have  ...
+
+As a result, it was agreed that a new component `LinkControl` would be created to realise the new hyperlink creation interface. This new UI would begin life as an experimental component which would consume `URLInput` internally. The API of `URLInput` would be enhanced as required with "experimental" features to facilitate the implementation of the new UI with the goal of eventually phasing
+out the use of `URLInput` entirely.
```
Unneeded line break.

jeryj commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -1,8 +1,49 @@
+
+## Search Suggestions
+
+Currently LinkControl will handle two types of input to create hyperlinks:
```
LinkControl should be `LinkControl`

jeryj commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -1,8 +1,49 @@
+Currently LinkControl will handle two types of input to create hyperlinks:
+
+1. Entity searches - the user may input text-based search queries for entities retrieved from remove data sources (in the context of WordPress these are `Pages`). For example a user might search for a `Page` they have just created by name (eg: About) and the UI will return a matching result if found.
+2. Direct entry - the user may also enter any arbitrary URL-like text. This includes full URLs (https://), URL fragements (eg: `#myinternallink`), `tel` protocol links (eg: `tel: 0800 1234`) and `mailto` protocol links (eg: `mailto: hello@wordpress.org).
```
`mailto: hello@wordpress.org`). is missing the closing apostrophe-ish mark to close the code comment.

jeryj commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -1,8 +1,49 @@
+
+### Data sources
+
+By default LinkControl utilizes the `__experimentalFetchLinkSuggestions` API from `core/block-editor` in order to retrieve search suggestions for matching `Page` post-type entities.
```
LinkControl should be `LinkControl`

jeryj commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -13,13 +54,13 @@ for how that link is expected to behave.
-- `opensInNewTab` (`boolean`, optional): Whether link should open in a new browser tab.This value is only assigned if not providing a custom `settings` prop.
+- `opensInNewTab` (`boolean`, optional): Whether link should open in a new browser tab. This value is only assigned if not providing a custom `settings` prop.
 
 ### settings
```
Should this be ### `settings` since it's a prop?

jeryj commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -35,7 +76,7 @@ Default properties include:
-An array of settings objects. Each object will used to render a `ToggleControl` for that setting.
+An array of settings objects associated with a link (for example: a setting to determine whether or not the link opens in a new tab). Each object will be used to render a `ToggleControl` for that setting.
 
 ### onClose
```
Should this be ### `onClose` since it's a prop?

jeryj commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -70,37 +111,41 @@ Whether to present initial suggestions immediately.
+If passed as either `true` or `false`, controls the internal editing state of the component to respectively show or not to show the URL input field.
 
 
 ### createSuggestion
```
Should this be ### `createSuggestion` since it's a prop?

jeryj commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -70,37 +111,41 @@ Whether to present initial suggestions immediately.
 
 - Type: `function`
 - Required: No
+- Returns: when called may return either a new `suggestion` directly or a `Promise` which resolves to a
```
Other comments have used a capital letter on the beginning of the comment, ie "Returns: When..." instead of "Returns: when..."

jeryj commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -70,37 +111,41 @@ Whether to present initial suggestions immediately.
-new suggestion. Suggestions have the following shape:
+This `suggestion` will then be _automatically_ passed to the `onChange` handler to create **the next link value**.
+
+As a result of the above, this prop is often used to allow on-the-fly creation of new entities (eg: `Posts`, `Pages`) based on the text the user has entered into the link search UI. As an example, the Navigation Block uses `createSuggestion` to create Pages on the fly from within the Block itself.
```
Other places have used "on-the-fly" instead of "on the fly". After Google-ing it, it appears all instances of "on-the-fly" should be "on the fly"

jeryj reviewed (changes requested):
Found more minor grammar and formatting things. Should be very quick fixes. Sorry I didn't catch all of those in the first review :/

jeryj committed:
Fixed unecessary line break

jeryj committed:
Fixed missing closing code comment apostrophe mark

jeryj committed:
Replaced instances of on-the-fly to on the fly

jeryj committed:
Fixed capitalization error

jeryj committed:
Wrapped all instances of LinkControl to  in the README

jeryj committed:
Fixed unnecessary line break

jeryj committed:
Added/Removed empty lines to be consisently spaced between heading levels

getdave commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -70,34 +110,33 @@ Whether to present initial suggestions immediately.
 
-If passed as either `true` or `false`, controls the internal editing state of the component to respective show or not show the URL input field.
-
+If passed as either `true` or `false`, controls the internal editing state of the component to respectively show or not to show the URL input field.
```
```suggestion
Controls the internal editing state of the component. If passed as either `true` or `false` will respectively show or hide the URL input field.
```
Rewording to make this clearer.

Also can we check whether it really has to be `true` or `false` as strict value types. For example if I pass `1` and it is cooerced into a boolean then does this cause some other unexpected behaviour? How about `undefined`?

getdave commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -70,34 +110,33 @@ Whether to present initial suggestions immediately.
-	url: // any string representing the URL value
+	id: // uniquely identifies the suggestion.
+	type: // the type of the suggestion (eg: `post`).
+	title: // Human-readable label for the suggestion.
```
```suggestion
	title: // human-readable label for the suggestion.
```

Consistent case

getdave reviewed (commented):
Noticed some other points that need addressing. If you'd like to do these @jeryj be my guest, otherwise I'll pick them up in the morning.

Thanks for your previous commits!

getdave committed:
Fix for consistent case

getdave committed:
Reword sentence to be easier to read.

getdave committed:
Remove legacy prop onClose

This was used when there was a Popover. There is no longer a Popover.

getdave commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -35,7 +75,7 @@ Default properties include:
-An array of settings objects. Each object will used to render a `ToggleControl` for that setting.
+An array of settings objects associated with a link (for example: a setting to determine whether or not the link opens in a new tab). Each object will be used to render a `ToggleControl` for that setting.
 
 ### onClose
```
Fixed.

getdave committed:
Fix missing space.

getdave committed:
Fix to clarify sentence subject.

getdave committed:
Update to indicate that entities searches are free text based.

getdave committed:
Improve clarity of description around how searches work

getdave committed:
Tweak to wording to improve clarity of sentence

getdave committed:
Removes implication that `URLInput` is going to be deprecated.

getdave committed:
Fix missing `a` to ensure sentence makes sense.

getdave committed:
Correct wording to be “remote”

getdave commented:
@aduth I ended up merging this as I think it's an improvement over the previous version. I'd welcome you taking a look and correcting any points where you don't agree.

aduth commented:
Thanks for this. I think it reads well. I don't know that I would have considered the following suggestions as blocking, but they would be good to consider for future revisions to this document:

- Some clear flagging/warning that this component is currently experimental. Technically speaking, experimental APIs are [not to be documented at all](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#experimental-and-unstable-apis), but I think documentation serves a crucial role in thinking through how this complex component evolves. Some leading warning messaging could be a sufficient compromise between these competing goals.
- Documenting an intentional correspondence between this component amongst its peers of "*Control" components (e.g. `TextControl`, `RadioControl`, etc).
- A simple usage example, including import snippet.
- Documenting intentions to backport some of these experimentations into the root URLInput component (some examples at https://github.com/WordPress/gutenberg/issues/18061#issuecomment-580480464).
